group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java

```diff
@@ -1073,6 +1151,314 @@ public CoordinatorShardBuilder<MockCoordinatorShard, String> get() {
+        } else {
+            // Or they are gone if aborted.
+            assertEquals(Collections.emptySet(), ctx.coordinator.coordinator().records());
+        }
+
+        // Records have been written to the log.
+        assertEquals(Arrays.asList(
+            InMemoryPartitionWriter.LogEntry.value(100L, (short) 5, "record1"),
+            InMemoryPartitionWriter.LogEntry.value(100L, (short) 5, "record2"),
+            InMemoryPartitionWriter.LogEntry.control(100L, (short) 5, 10, result)
+        ), writer.entries(TP));
+
+        // Commit write #1.
+        writer.commit(TP, 2);
+
+        // The write is completed.
+        assertTrue(write1.isDone());
+        assertEquals("response1", write1.get(5, TimeUnit.SECONDS));
+
+        // Commit completion #1.
+        writer.commit(TP, 3);
+
+        // The transaction is completed.
+        assertTrue(complete1.isDone());
+        assertNull(complete1.get(5, TimeUnit.SECONDS));
+    }
+
+    @Test
+    public void testScheduleTransactionCompletionWhenWriteTimesOut() throws InterruptedException {
+        MockTimer timer = new MockTimer();
+        MockPartitionWriter writer = new MockPartitionWriter();
+
+        CoordinatorRuntime<MockCoordinatorShard, String> runtime =
+            new CoordinatorRuntime.Builder<MockCoordinatorShard, String>()
+                .withTime(timer.time())
+                .withTimer(timer)
+                .withDefaultWriteTimeOut(DEFAULT_WRITE_TIMEOUT)
+                .withLoader(new MockCoordinatorLoader())
+                .withEventProcessor(new DirectEventProcessor())
+                .withPartitionWriter(writer)
+                .withCoordinatorShardBuilderSupplier(new MockCoordinatorShardBuilderSupplier())
+                .withCoordinatorRuntimeMetrics(mock(GroupCoordinatorRuntimeMetrics.class))
+                .withCoordinatorMetrics(mock(GroupCoordinatorMetrics.class))
+                .build();
+
+        // Loads the coordinator.
+        runtime.scheduleLoadOperation(TP, 10);
+
+        // Verify the initial state.
+        CoordinatorRuntime<MockCoordinatorShard, String>.CoordinatorContext ctx = runtime.contextOrThrow(TP);
+        assertEquals(0, ctx.coordinator.lastWrittenOffset());
+        assertEquals(0, ctx.coordinator.lastCommittedOffset());
+        assertEquals(Collections.singletonList(0L), ctx.coordinator.snapshotRegistry().epochsList());
+
+        // Complete #1. We should get a TimeoutException because the HWM will not advance.
+        CompletableFuture<Void> timedOutCompletion = runtime.scheduleTransactionCompletion(
+            "complete#1",
+            TP,
+            100L,
+            (short) 5,
+            10,
+            TransactionResult.COMMIT,
+            Duration.ofMillis(3)
+        );
+
+        timer.advanceClock(4);
+
+        assertFutureThrows(timedOutCompletion, org.apache.kafka.common.errors.TimeoutException.class);
```

Comment:
should we ensure pending is emptied, we restore the previous offset etc.?

Reply:
Sure. I can add validation. However, note that we don't revert to the previous state in this case because the operation has been written to the log so we cannot revert anymore. We only timeout the operation if the operation is not committed (based on HWM) before the timeout is reached.

Reply:
Hmmm -- I guess I thought the timeout case is the case we revert. Which case is the case we revert?

Reply:
We revert in two [cases](https://github.com/apache/kafka/pull/14985/files#diff-4b9dd21d7748bfb375634ef6e39037e99f53ed2546519bec32fa1707bd262fe8): 1) The update cannot be applied to the state machine; 2) The update cannot be written to the partition/log. By this, I mean that the write fails directly (e.g. not leader, batch too large, etc). If the change accepted by the leader, we cannot revert the state because it would not be consistent.

Reply:
synced offline here. We only rollback if the write fails. If the write completes we either wait for it to replicate or lose the info if the leader changes.